Fix the advanced features section of distinctive lang features (#3673)
ballerina-platform/ballerina-dev-website: learn/guides/distinctive-language-features/advanced-general-purpose-language-features.md

description: Let's now look at the other features of the Ballerina language. These are a mixed bag of additional options to the language, making everything fit together to build a Ballerina application.
keywords: ballerina, programming language, ballerina packages,language-guide
permalink: /learn/distinctive-language-features/advanced-general-purpose-language-features/
active: advanced-general-purpose-language-features
intro: Let's now look at the other features of the Ballerina language. These are a mixed bag of additional options to the language, making everything fit together to build a Ballerina application.
redirect_from:
- /learn/distinctive-language-features/advanced-general-purpose-language-features
---

## Default Values for Function Parameters

Ballerina allows specifying default values for function parameters. You can use any expression such as a literal or a function call as the default value of a parameter. Additionally, the default value expressions can use the values of preceding parameters.

```ballerina
function substring(string str, int 'start = 0, int end = str.length()) 
            returns int? {
    // ...
}
```

> The way it works is, the type descriptor contains closures for every defaultable parameter, which has the values for previous parameters as parameters for those closures. Using those closures, the compiler generates code to fill missing values at the call site for this function. This is not part of the function type and is not applicable when functions are passed around as a first-class value.

Author: ```suggestion
The type descriptor contains closures for every defaultable parameter. These closures accept the values specified for the previous parameters as arguments. Using those closures, the compiler generates code to fill missing values at the call site for this function. This is not part of the function type and is not applicable when functions are passed around as a first-class value.
```

## Providing Function Arguments by Name

> In addition to default values for parameters, it is also useful to call functions with argument names. Ballerina allows you to call functions where arguments can be supplied by name.

Reviewer: We can emphasize parameters instead of calling.
`We can provide values to parameters in any order using named arguments`

> In addition to default values for parameters, it is also useful to call functions with argument names. Ballerina allows you to call functions where arguments can be supplied by name.

Reviewer: "named arguments" is a known concept in programming languages.
```suggestion
In addition to default values for parameters, it is also useful to call functions with named arguments.
```

Author: Incorporating both comments

```suggestion
In addition to default values for parameters, it is also useful to call functions with named arguments. Ballerina allows you to call functions with named arguments, and these named arguments do not have to be in the same order as the parameters.
```

Consider the function below.

```ballerina
function foo(int x, int y, int z) {
    
}
```

This function can be called in any of the following ways:

```ballerina
foo(1, 2, 3);
foo(x = 1, y = 2, z = 3);
foo(z = 3, y = 2, x = 1);
foo(1, z = 3, y = 2);
```

> The first option calls the function in the usual way without using the names of the arguments. The second and third options use named arguments that are transformed into positional arguments by the Ballerina compiler. The argument list of a function is described by a tuple type, and names are not part of that type. You can also combine the named arguments with unnamed ones, as shown in the fourth option.

Reviewer: ```suggestion
The first statement calls the function in the usual way leveraging order of the parameters. The second and third statements use named arguments to provide values for parameters in different order. The argument list of a function is described by a tuple type, and names are not part of that type. You can also provide named arguments after positional arguments in any order, as shown in the fourth statement.
```

Author: ```suggestion
The first statement calls the function in the usual way without using the names of the arguments. The second and third statements use named arguments in an order different from that of the parameters of the function definition. The Ballerina compiler transforms them into positional arguments. The argument list of a function is described by a tuple type, and names are not part of that type. You can also combine the named arguments with positional arguments, as shown in the fourth option, given that they are specified after the positional arguments.
```

This way of including record fields as named arguments in a function provides a consistent experience to the caller.

## Default Values for Record Fields

Ballerina also allows default values for record fields as part of the record's type descriptor. A default value is an expression, which can also be a closure that is computed in the context of the enclosing module scope.

```ballerina
type X record {
    string str = "";
};

X x = {};
```
  
In the above code example, the record **``X``** has a field **``str``** which is defaulted to empty string. Since the variable **``x``** of type **``X``** is initialized without specifying a value for the ``str`` field, the constructed ``X`` value will use the default empty string value for the ``str`` field.

Default values do not affect static typing. They only affect the use of type descriptors to construct records. Calling the ``value:cloneWithType()`` function with a record type-descriptor **``T``** will make use of default values in **``T``** if required. Similarly, using **``*T``** also copies the default values.

## Object Types

> Ballerina also allows you to define object types that define just the type without the implementation. Unlike classes, which give you the type as well as the implementation that you can instantiate through ``new``, an object type is a definition without any implementation. It is similar to the concept of an interface, as defined in the Java programming language.

Author: I feel like "as well as" emphasizes both type and implementation a bit more.

```suggestion
Ballerina also allows you to define object types that define just the type without the implementation. Unlike classes, which give you the type as well as the implementation that you can instantiate through ``new``, an object type is a definition without any implementation. It is similar to an interface, as defined in the Java programming language.
```

```ballerina
type Cloneable object {
    function clone() returns Cloneable;
};

type Shape object {
    *Cloneable;
    
    function draw();
};

class Circle {
    *Shape;

    function clone() returns Circle {
        return new;
    }

    function draw() {
    }
}
```

In the above code example, **``Cloneable``** is an object type. It is included as part of the interface for the **``Shape``** object type. Therefore **``Shape``** defines an interface containing the function **``draw()``** which is part of its own type as well as the function **``clone()``** included from the **``Cloneable``** type. The class **``Circle``** includes the **``Shape``** object type. Therefore it has to implement both the **``clone()``** and **``draw()``** methods. The function **``clone( )``** returns a **``Circle``** and is considered valid since the **``Circle``** type becomes a subtype of **``Cloneable``** type once it includes the **``Cloneable``** object type.  

The implementation of the object type within the class that includes the type is checked at the compile time. This provides interface inheritance. Ballerina does not support implementation inheritance.

## Distinct Object Types

> Ballerina also supports the concept of ``distinct`` object types. This allows you to define a type with a name that is significant, similar to nominal typing, within a structured type system. It is done by using the ``distinct`` keyword in the type definition.

Reviewer: ```suggestion
Ballerina also supports the concept of ``distinct`` object types. This concept allows you to define a type with a name that is significant, similar to nominal typing, within a structured type system. It is done using the ``distinct`` keyword in the type definition.
```

Reviewer: "It is done" is passive voice

distinct class Employee {
    *Person;
    
    function init(string name) {
        self.name = name;
    }
}

distinct class Manager {
    *Person;
    
    function init(string name) {
        self.name = name;
    }
}
```

> In the above code example, the **``Person``** type is defined as ``distinct``. This means that the compiler will give this type a unique name, with which the object value will be tagged and is unique across all the distinct types defined in the program. The **``Employee``** type includes the **``Person``** type, so it belongs to the distinct **``Person``** type as well as the distinct **``Employee``** type. Similarly, the **``Manager``** class also inherits the **``Person``** type.

Reviewer: ```suggestion
In the above code example, the **``Person``** type is defined as ``distinct``. This means that the compiler will give this type a unique name, with which the object value will be tagged and is unique across all the distinct types defined in the program. The **``Employee``** type includes the **``Person``** type, so it belongs to the distinct **``Person``** type and the distinct **``Employee``** type. Similarly, the **``Manager``** class also inherits the **``Person``** type.
```

type FooBar Foo & Bar;

// Same as
type FooBar object {
    *Foo;
    *Bar;
};
```

In the above code example, **``Foo``** and **``Bar``** are two separate object types. The **``FooBar``** type is defined using the type intersection of **``Foo``** and **``Bar``** using the ``&`` notation. This has the same result as including the types using the ``*`` notation. It provides a neat alternative to type inclusion but is less flexible.

## Expression-Oriented Style

> Ballerina's general philosophy is to make the programming syntax familiar to those familiar with the C-family of languages. It has a distinction between statements and expressions, just like in C. This is the imperative style of programming which is considered less intuitive. That's why there is a shift towards a functional style of programming that is more expression-oriented.

Reviewer: ```suggestion
Ballerina's general philosophy is to make the programming syntax familiar to those familiar with the C-family of languages. This is the imperative style of programming which is considered less intuitive. It has a distinction between statements and expressions, just like in C. That's why there is a shift towards a functional programming style that is more expression-oriented.
```

Author: Is switching the sentences correct here?

Reviewer: Lets keep

Reviewer: Lets keep

Reviewer: Lets keep

## Tuples

> The tuple type is another structured type supported by Ballerina, which creates a list of values just like arrays. The main difference between arrays and tuples is that an array has only one type applicable to every member of its list, whereas in a tuple type you can individually specify the types for each member.

Reviewer: ```suggestion
The tuple type is another structured type supported by Ballerina, which creates a list of values like arrays. The main difference between arrays and tuples is that an array has only one type applicable to every member of its list. In contrast, in a tuple type, you can individually specify the types for each member.
```
